source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,272 @@
+
+    int getPriority();
+
+    boolean supports(DataHolder dataHolder);
+
+    Optional<M> from(DataHolder dataHolder);
+
+    Optional<M> createFrom(DataHolder dataHolder);
+
+    Optional<M> fill(DataHolder dataHolder, M manipulator, MergeFunction overlap);
+
+    Optional<M> fill(DataContainer container, M m);
+
+    DataTransactionResult set(DataHolder dataHolder, M manipulator, MergeFunction function);
+
+    Optional<I> with(Key<? extends BaseValue<?>> key, Object value, I immutable);
+
+    DataTransactionResult remove(DataHolder dataHolder);
+ }
+
+getPriority() method
+--------------------
+
+The first thing that we notice is a ``getPriority()`` method. This is used for multi-processor registration, or in more
+plain terms: A multi-processor registration system for mods to provide their own compatibility processors for their
+custom entity/itemstack/tileentity/blockstate types. The higher the priority return, the `earlier` the processor is used
+in what's called the ``DataProcessorDelegate``. The delegate is a collection of registered ``DataProcessor`` of a
+particular ``DataManipulator`` such that all registered ``DataProcessor`` can be used.
+
+supports(DataHolder)
+--------------------
+
+Next up, we have ``supports(DataHolder)``. This is really simple, if your ``DataManipulator`` is supposed to be
+"supported" by the provided ``DataHolder`` (which can be an ItemStack, Entity, TileEntity, etc.) you can define the
+"support" in this method. So, if I have a ``VelocityDataProcessor``, my ``support(DataHolder)`` would look something like so:
+
+.. code-block:: java
+
+ @Override
+ public boolean supports(DataHolder dataHolder) {
+  return dataHolder instanceof Entity;
+ }
+
+from(DataHolder)
+----------------
+
+createFrom(DataHolder)
+----------------------
+
+.. warning::
+
+ Not sure the code is right. Please doublecheck!
+ Code taken out of PR#139 for now.
+
+.. code-block:: java
+
+ public Optional<FlyingData> createFrom(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            boolean isFlying = ((net.minecraft.entity.Entity) dataHolder).isAirBorne;
+            if (dataHolder instanceof EntityPlayer) {
+                isFlying = ((EntityPlayer) dataHolder).capabilities.allowFlying;
+            }
+            return Optional.<FlyingData>of(new SpongeFlyingData(isFlying));
+        }
+        return Optional.absent();
+    }
+
+fill(DataHolder)
+----------------
+
+.. warning::
+
+  Not sure the code is right, especially the ``checkNotNull`` part. Please doublecheck!
+  Code taken out of PR#139 for now.
+
+.. code-block:: java
+
+    public Optional<FlyingData> fill(DataHolder dataHolder, FlyingData manipulator, MergeFunction overlap) {
+        if (supports(dataHolder)) {
+            final FlyingData merged = checkNotNull(overlap.merge(checkNotNull(manipulator).copy(), from(dataHolder).get()), "MergeFunction cannot be null!");
```

Comment:
The line is suppoed to read ` final FlyingData merged = checkNotNull(overlap, "MergeFunction cannot be null!").merge(checkNotNull(manipulator).copy(), from(dataHolder).orNull()));